src/store/reducers.js

```diff
@@ -1,58 +1,110 @@
+        {},
+        currentSession,
+        payload // login , password
+      )
+    case actions.SIGN_IN :
+      // user tried to sign in
+      return Object.assign(
+        {},
+        currentSession,
+        action.payload, // login , password
+        {isLoggingIn: true}
+      )
 
-const createAsyncHandler = ({ error, next }) => (state, action) => {
-  if (action.error) {
-    if (error) {
-      return error(state, action.payload)
-    }
-  } else {
-    if (next) {
-      return next(state, action.payload)
-    }
-  }
-
-  return state
-}
+    case actions.SIGN_OUT :
+      return Object.assign({}, currentSession, {isLoggingOut: true})
 
-// Action handlers are reducers but bound to a specific action.
-const combineActionHandlers = (initialState, handlers) => {
-  for (const action in handlers) {
-    const handler = handlers[action]
-    if (typeof handler === 'object') {
-      handlers[action] = createAsyncHandler(handler)
-    }
-  }
+    case actions.SIGNED_OUT:
+      // when user is signed out : reset session
+      return initialState.session
 
-  return (state = initialState, action) => {
-    const handler = handlers[action.type]
+    case actions.SIGNED_IN :
+      // when user is signed in : session informations are received from server
+      return Object.assign(
+        {},
+        currentSession,
+        action.payload,
+        {
+          isLoggued: true,
+          isLoggingIn: false,
+          isLoggingOut: false,
+          password: 'redacted , don t need it '
+        }
+      )
 
-    return handler
-      ? handler(state, action)
-      : state
+    default :
+      return currentSession // reducer HAVE TO return a state, even when they did nothing
   }
 }
 
-// ===================================================================
+function xoApiReducer (state = initialState.xoApi, action) {
+  let payload = action.payload || {}
+  switch (action.type){
+    case actions.VM_EDIT:
+      payload = Object.assign({},
+        state[action.payload.id],
+        payload,
+        {isSaved: false}
+      )
+      return Object.assign(
+        {},
+        state,
+        {
+          [action.payload.id]: payload
+        }
+      )
 
-export const counter = combineActionHandlers(0, {
-  [actions.decrement]: counter => counter - 1,
-  [actions.increment]: counter => counter + 1
-})
+    case actions.VM_SAVE:
+    console.log(' SAVE ',action.payload)
+      return Object.assign(
+        {},
+        state,
+        {
+          [action.payload.id]: {isSaving:true}
+        }
+      )
+      return state
+    case actions.VM_SAVED:
+      // put the VM at its real id if it's a creation
+      payload = Object.assign({},
+        state[action.payload.id],
+        payload,
+        {isSaved: true, isSaving: false}
+      )
+      console.log('VM_SAVED',payload)
+      return Object.assign(
+        {},
+        state,
+        {
+          [action.payload.id]: payload
+        }
+      )
+  }
+  return state
+}
```

Comment:
I don't think we should use such a big reducers, it's going to get enormous as we add new functionalities.